### ISSUE TITLE
SSL Handshake errors after upgrading to 24.5.1.1763

### ISSUE THREAD

**Issue by joshbartley** [ClickHouse/ClickHouse#64735]
**Describe what's wrong**
3-node cluster and after upgrading 2 nodes to 24.5.1.1763 from 23.7.1.2470, haproxy started reporting a L6RSP (SSL Handshake failure). Certificates are publicly signed, not expired, and working configs for the past 9 months. 

Chrome also reports a SSL_PROTOCOL_ERROR on occasion when connecting directly to a node but it also keeps a connection alive unlike haproxy so it's harder to reproduce.

openssl from the load balancer reports everything as fine.
`openssl s_client -connect _myhoustname_:8443 < /dev/null`
![image](https://github.com/ClickHouse/ClickHouse/assets/5738532/d70db724-7a7a-4418-94fe-bdc8f696aaf7)

**Does it reproduce on the most recent release?**
Yes

**Enable crash reporting**
Not crashing.

**How to reproduce**
Honestly, no idea. I don't think our setup is that unique as we followed the SSL-TLS Guide

**Expected behavior**
No handshake errors?

**Error message and/or stacktrace**
I found this around the times that was a handshake error. It's pretty frequent but the clickhouse servers don't appear to have an issue talking to each other.
```
2024.06.02 20:00:19.335772 [ 8478 ] {} <Error> ServerErrorHandler: Code: 210. DB::NetException: SSL Exception: error:0A000115:SSL routines::session id context uninitialized, while reading from socket (peer: 172.18.11.105:40784, local: 172.18.11.107:8443). (NETWORK_ERROR), Stack trace (when copying this message, always include the lines below):

0. DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000c5c4fbb
1. DB::NetException::NetException<String, String, String>(int, FormatStringHelperImpl<std::type_identity<String>::type, std::type_identity<String>::type, std::type_identity<String>::type>, String&&, String&&, String&&) @ 0x000000000c6ff8be
2. DB::ReadBufferFromPocoSocket::nextImpl() @ 0x000000000c6ff39a
3. DB::HTTPServerRequest::HTTPServerRequest(std::shared_ptr<DB::IHTTPContext>, DB::HTTPServerResponse&, Poco::Net::HTTPServerSession&, StrongTypedef<unsigned long, ProfileEvents::EventTag> const&) @ 0x0000000011ea76c2
4. DB::HTTPServerConnection::run() @ 0x0000000011ea5ef8
5. Poco::Net::TCPServerConnection::start() @ 0x0000000014768447
6. Poco::Net::TCPServerDispatcher::run() @ 0x00000000147688d9
7. Poco::PooledThread::run() @ 0x000000001485ed41
8. Poco::ThreadImpl::runnableEntry(void*) @ 0x000000001485d31d
9. ? @ 0x00007fbe57780ac3
10. ? @ 0x00007fbe57812850
 (version 24.5.1.1763 (official build))
```

> If applicable, add screenshots to help explain your problem.

**Additional context**
Fully patched ubuntu servers (22.04.4 LTS)
Certificates are publicly signed, not expired, and are wildcard certificates.

Haproxy ssl config
```
ssl-default-bind-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        ssl-default-bind-options prefer-client-ciphers no-sslv3 no-tlsv10 no-tlsv11 no-tlsv12 no-tls-tickets

        ssl-default-server-ciphersuites TLS_AES_128_GCM_SHA256:TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
        ssl-default-server-options no-sslv3 no-tlsv10 no-tlsv11 no-tlsv12 no-tls-tickets
```

Clickhouse openssl config. I tried both `cachesession true and false with the same result, I was desperate.
```
<clickhouse>
  <openSSL>
    <server>
        <loadDefaultCAFile>true</loadDefaultCAFile>
        <certificateFile>/etc/clickhouse-server/certs/****************.crt</certificateFile>
        <privateKeyFile>/etc/clickhouse-server/certs/************.key</privateKeyFile>
        <verificationMode>relaxed</verificationMode>
        <caConfig>/etc/clickhouse-server/certs/*********_ca.crt</caConfig>
        <cacheSessions>true</cacheSessions>
        <disableProtocols>sslv2,sslv3</disableProtocols>
        <preferServerCiphers>true</preferServerCiphers>
    </server>
    <client>
        <loadDefaultCAFile>false</loadDefaultCAFile>
        <caConfig>/etc/clickhouse-server/certs/********_ca.crt</caConfig>
        <cacheSessions>true</cacheSessions>
        <disableProtocols>sslv2,sslv3</disableProtocols>
        <preferServerCiphers>true</preferServerCiphers>
        <verificationMode>relaxed</verificationMode>
        <invalidCertificateHandler>
                        <name>RejectCertificateHandler</name>
        </invalidCertificateHandler>
     </client>
  </openSSL>
</clickhouse>
```

with cacheSession false

```
<clickhouse>
  <openSSL>
    <server>
        <loadDefaultCAFile>true</loadDefaultCAFile>
        <certificateFile>/etc/clickhouse-server/certs/****************.crt</certificateFile>
        <privateKeyFile>/etc/clickhouse-server/certs/************.key</privateKeyFile>
        <verificationMode>relaxed</verificationMode>
        <caConfig>/etc/clickhouse-server/certs/*********_ca.crt</caConfig>
        <cacheSessions>false</cacheSessions>
        <disableProtocols>sslv2,sslv3</disableProtocols>
        <preferServerCiphers>true</preferServerCiphers>
    </server>
    <client>
        <loadDefaultCAFile>false</loadDefaultCAFile>
        <caConfig>/etc/clickhouse-server/certs/********_ca.crt</caConfig>
        <cacheSessions>true</cacheSessions>
        <disableProtocols>sslv2,sslv3</disableProtocols>
        <preferServerCiphers>false</preferServerCiphers>
        <verificationMode>relaxed</verificationMode>
        <invalidCertificateHandler>
                        <name>RejectCertificateHandler</name>
        </invalidCertificateHandler>
     </client>
  </openSSL>
</clickhouse>
```

**Comment by joshbartley** [comment]
A dotnet client also reports sporadic SSL errors after upgrading if it's pointed directly to the server's https port.

**Comment by joshbartley** [comment]
Is it related to this commit?

https://github.com/ClickHouse/ClickHouse/pull/59870/commits/d30b48f54e42ec5aeb1bc27c937adb8efdb8108d

**Comment by xogoodnow** [comment]
Hi @joshbartley,
Could you test it with <cacheSessions>false</cacheSessions>?
for both the client and server

I have just updated to 24.5.1.1763 from 24.3 and did not see this error (only on lab, I have not tested the cluster under heavy load yet)

**Comment by xogoodnow** [comment]
Yes set the cachesession to false and see if the problem persists.
Also I would recommend to omit the part about <preferServerCiphers>
My approach is, when you do not know what is causing the problem, remove as much variables as possible.

Also my clients do not directly access Clickhouse, I'm using self signed certificates.
Also I do not terminate ssl on ha-Proxy, the client that you are using within your application/api can have the health check and there would be no need for ha-Proxy.

But since we are considering upgrading to the newest version, testing this matter would be crucial for us.

**Comment by joshbartley** [comment]
`cacheSession` is false and the issue still occurs on 24.5.1.1763. 

![image](https://github.com/ClickHouse/ClickHouse/assets/5738532/07467f4a-c196-4b66-a66f-1594045c766c)

Error still also occurs with a C# client connecting directly to the node.

**Comment by xogoodnow** [comment]
It's just guys, but could you remove the disabled protocols section as well, your c client might not accept the cipher suites defined by ha-Proxy and Clickhouse.
If it is possible, relax the ssl protocols and ciphers and let them find the best cipher suite among themselves instead of explicitly defining them.
Still I do not get this problem (we are using Go client"

**Comment by joshbartley** [comment]
All the nodes are upgraded to the same version. I'm not enabling sslv3, or a bad version of TLS. Everything was working before the upgrading just fine.

**Comment by xogoodnow** [comment]
Hi @joshbartley, oh of course, if your cluster in deployed on production.

I still have not encountered this issue, sorry for not being able to help

**Comment by bjne** [comment]
I had to downgrade all the way to 24.3.3.102 to resolve this issue

**Comment by xogoodnow** [comment]
We have halted the upgrade as well,
currently using, v24.3.2.23-lts

**Comment by IngaFeick** [comment]
We encountered the same issue when upgrading from 24.1.8.22 to 24.5.1.1763. We tried removing the disabledProtocols to no effect. Rolling back to 24.3.3.102 as well.

**Comment by joshbartley** [comment]
Looks like it's not just me. Updating the thread with the Clickhouse Cloud announcement for the on-prem people.

![image](https://github.com/ClickHouse/ClickHouse/assets/5738532/952c63c5-5a2a-4134-a50d-cf3e7f69d223)

**Comment by rschu1ze** [comment]
We indeed hit this issue internally and it seems hard to reproduce :-/

I think it is caused by the disabled session caching (https://github.com/ClickHouse/ClickHouse/pull/59870/commits/d30b48f54e42ec5aeb1bc27c937adb8efdb8108d - also mentioned in above discussion). Turning session caching on or off via configuration file (`<cacheSessions>`) currently does nothing.

Session caching is re-enabled here: https://github.com/ClickHouse/ClickHouse/pull/65111 ... unfortunately that uncovered another issue in OpenSSL: https://github.com/openssl/openssl/issues/24629. For the time being I made a temporary fix as part of the session caching re-enablement (--> https://github.com/ClickHouse/ClickHouse/pull/65111/commits/7abc4a5a8c32a58f15cf794a522b1e46fdfa9042).

It would be great if anyone could pick up a build of https://github.com/ClickHouse/ClickHouse/pull/65111, run it and check if it resolves the problem. In parallel, I'll also try reproducing.

**Comment by rschu1ze** [comment]
@joshbartley Builds can be fetched from continuous integration (CI). I just changed something in the PR, so CI restarts from scratch. Builds are expected to appear 1-2 hours after the last changes in the PR.

Basically, go [to this comment](https://github.com/ClickHouse/ClickHouse/pull/65111#issuecomment-2161112389).

Then click "Click here". This opens an overview page of all completed CI jobs.

Once the build job is finished, the overview page shows something like this:

![image](https://github.com/ClickHouse/ClickHouse/assets/100446237/5875862d-73a3-4814-8f7c-bd03c6986113)

Click "report.hmtl" and search "package_release" (multiple entries exist). The right column provides links to different installation options, e.g. deb packages, rpm packages or plain tgz archives. The installation options are the same as the ones on the release page: https://github.com/ClickHouse/ClickHouse/releases

**Comment by rschu1ze** [comment]
After some local testing using an internal repro, I am 99% sure that #65111 resolves the issue.
